embassy/src/executor/raw/mod.rs

```diff
@@ -207,6 +211,148 @@ impl<F: Future + 'static> TaskStorage<F> {
 
 unsafe impl<F: Future + 'static> Sync for TaskStorage<F> {}
 
+/// TODO
+
+// repr(C) is needed to guarantee that the Task is located at offset 0
+// This makes it safe to cast between TaskHeader and TaskStorage pointers.
+#[repr(C)]
+pub struct ScopedTaskStorage<T> {
+    header: TaskHeader,
+    future: UninitCell<*mut dyn Future<Output = T>>, // Valid if STATE_SPAWNED
+    result: UnsafeCell<MaybeUninit<T>>,              // Valid if STATE_RESULT_PENDING
+    waker: AtomicWaker,
+}
+
+impl<T> ScopedTaskStorage<T> {
+    /// Create a new TaskStorage, in not-spawned state.
+    #[cfg(feature = "nightly")]
+    pub const fn new() -> Self {
+        Self {
+            header: TaskHeader::new(),
+            future: UninitCell::uninit(),
+            result: UnsafeCell::new(MaybeUninit::uninit()),
+            waker: AtomicWaker::new(),
+        }
+    }
+
+    /// Create a new TaskStorage, in not-spawned state.
+    #[cfg(not(feature = "nightly"))]
+    pub fn new() -> Self {
+        Self {
+            header: TaskHeader::new(),
+            future: UninitCell::uninit(),
+            result: UnsafeCell::new(MaybeUninit::uninit()),
+            waker: AtomicWaker::new(),
+        }
+    }
+
+    /// TODO
+    pub fn spawn_scoped<'a>(
+        &'static self,
+        future: &'a mut (dyn Future<Output = T> + 'a),
+    ) -> Option<(SpawnScopedGuard<'a, T>, SpawnToken<impl Sized + 'a>)> {
+        if self.spawn_mark_used() {
+            let guard = SpawnScopedGuard { storage: self };
+            let token = unsafe {
+                SpawnToken::<&'a mut (dyn Future<Output = T> + 'a)>::new(
+                    self.spawn_initialize(core::mem::transmute(future as *mut dyn Future<Output = T>)),
+                )
+            };
+
+            return Some((guard, token));
+        }
+
+        None
+    }
+
+    fn spawn_mark_used(&self) -> bool {
+        let state = STATE_SPAWNED | STATE_RUN_QUEUED;
+        self.header
+            .state
+            .compare_exchange(0, state, Ordering::AcqRel, Ordering::Acquire)
+            .is_ok()
+    }
+
+    unsafe fn spawn_initialize(&'static self, future: *mut dyn Future<Output = T>) -> NonNull<TaskHeader> {
+        // Initialize the task
+        self.header.poll_fn.write(Self::poll);
+        self.future.write(future);
+        NonNull::new_unchecked(&self.header as *const TaskHeader as *mut TaskHeader)
+    }
+
+    unsafe fn poll(p: NonNull<TaskHeader>) {
+        let this = &*(p.as_ptr() as *const Self);
+
+        let future = Pin::new_unchecked(&mut **this.future.as_mut());
+        let waker = waker::from_task(p);
+        let mut cx = Context::from_waker(&waker);
+        match future.poll(&mut cx) {
+            Poll::Ready(res) => {
+                this.future.drop_in_place();
+                this.header.state.fetch_and(!STATE_SPAWNED, Ordering::AcqRel);
+                (*this.result.get()).write(res);
```

Comment:
there's a tiny window here where `state=0`, where another thread could claim this TaskStorage even if we're not done with it.